src/prelude/internals.mo

```diff
@@ -448,140 +448,138 @@ func @call_raw(p : Principal, m : Text, a : Blob) : async Blob {
+    return
+  };
 
-    var gathered = 0;
-    let thunks = Array_init<?(() -> async ())>(10, null); // we want max 10
-
-    func gatherExpired(n : ?@Node) = switch n {
-        case null ();
-        case (?n) {
-            gatherExpired(n.pre);
-            if (n.expire[0] > 0 and n.expire[0] <= now and gathered < thunks.size()) {
-                thunks[gathered] := ?(n.job);
-                switch (n.delay) {
-                    case (?delay) {
-                        // re-add the node
-                        let expire = n.expire[0] + delay;
-                        // N.B. insert only works on pruned nodes
-                        func reinsert(m : ?@Node) : @Node = switch m {
-                            case null ({ n with expire = [var expire]; pre = null; post = null });
-                            case (?m) {
-                                assert m.expire[0] != 0;
-                                if (expire < m.expire[0]) ({ m with pre = ?reinsert(m.pre) })
-                                else ({ m with post = ?reinsert(m.post) })
-                            }
-                        };
-                        @timers := ?reinsert(@prune(@timers));
-                    };
-                    case _ ()
-                };
-                n.expire[0] := 0;
-                gathered += 1;
-            };
-            gatherExpired(n.post);
-        }
-    };
+  var gathered = 0;
+  let thunks = Array_init<?(() -> async ())>(10, null); // we want max 10
 
-    gatherExpired(@timers);
+  func gatherExpired(n : ?@Node) = switch n {
+    case null ();
+    case (?n) {
+      gatherExpired(n.pre);
+      if (n.expire[0] > 0 and n.expire[0] <= now and gathered < thunks.size()) {
+        thunks[gathered] := ?(n.job);
+        switch (n.delay) {
+          case (?delay) if (delay != 0) {
+            // re-add the node
+            let expire = n.expire[0] + delay;
+            // N.B. insert only works on pruned nodes
+            func reinsert(m : ?@Node) : @Node = switch m {
+              case null ({ n with expire = [var expire]; pre = null; post = null });
+              case (?m) {
+                assert m.expire[0] != 0;
+                if (expire < m.expire[0]) ({ m with pre = ?reinsert(m.pre) })
+                else ({ m with post = ?reinsert(m.post) })
+              }
+            };
+            @timers := ?reinsert(@prune(@timers));
+          };
+          case _ ()
+        };
+        n.expire[0] := 0;
+        gathered += 1;
+      };
+      gatherExpired(n.post);
+    }
+  };
 
-    let futures = Array_init<?(async ())>(thunks.size(), null);
-    for (k in thunks.keys()) {
-        futures[k] := switch (thunks[k]) { case (?thunk) ?thunk(); case _ null };
-    };
+  gatherExpired(@timers);
 
-    for (f in futures.vals()) {
-        switch f { case (?f) try await f catch _ { }; case _ () }
-    };
+  for (k in thunks.keys()) {
+    ignore switch (thunks[k]) { case (?thunk) ?thunk(); case _ null };
+  }
```

Comment:
```suggestion
  for (o in thunks.keys()) {
    switch (o) { case (?thunk) { ignore ?thunk()} ; case _ {} };
  }
```
(if you keep the array, this might be sweeter).

Reply:
I wonder what will happen with upgrades if the timers haven't been stopped. All of these async calls will extend the continuation table until completed, if we don't await them.

Ideally, motoko would offer a way to do oneway scheduling of async blocks, or local oneway functions to avoid setting up the continuations, but we don't have that facility yet.

Reply:
IIRC, the spec says that the global timer gets reset to 0 on upgrades.